dll/win32/iphlpapi/icmp.c

```diff
@@ -1,964 +1,339 @@
+        SetLastError(ERROR_NOT_ENOUGH_MEMORY);
         return 0;
     }
-    strcpy(ntoa_buffer, ptr);
-    argv[argc++] = ntoa_buffer;
-    argv[argc] = NULL;
 
-    /* Dump commandline for debugging purposes */
-    TRACE("Ping commandline: ");
-    for (i = 0; i < argc; i++)
+    /* Get the start of the dynamic data */
+    Ptr = (PBYTE)(IcmpEchoData + 1);
+
+    /* Fill out the struct */
+    IcmpEchoData->DestinationAddress = DestinationAddress;
+    IcmpEchoData->Timeout = Timeout;
+    IcmpEchoData->Ttl = RequestOptions->Ttl;
+    IcmpEchoData->Tos = RequestOptions->Tos;
+    IcmpEchoData->Flags = RequestOptions->Flags;
+    IcmpEchoData->OptionsSize = RequestOptions->OptionsSize;
+    if (RequestOptions->OptionsSize)
     {
-        TRACE("%s ", debugstr_a(argv[i]));
+        IcmpEchoData->OptionsData = Ptr;
+        CopyMemory(IcmpEchoData->OptionsData, RequestOptions->OptionsData, RequestOptions->OptionsSize);
+        Ptr += RequestOptions->OptionsSize;
     }
-    TRACE("\n");
-
-    /* Prefill the reply struct with fallback values */
-    memset(reply, 0, sizeof(*reply));
-    reply->Address       = DestinationAddress;
-    reply->RoundTripTime = 40;
-    reply->Options.Ttl   = 120;
-
-    /* Create communication pipes */
-#ifdef HAVE_PIPE2
-    if (pipe2(pipe_out, O_CLOEXEC) < 0)
-#endif
+    if (RequestSize)
+    {
+        IcmpEchoData->RequestData = Ptr;
+        CopyMemory(IcmpEchoData->RequestData, RequestData, RequestSize);
+    }
+
+    /* If we have no event or APC, then we'll need an internal wait object */
+    if ((Event == NULL) && (ApcRoutine == NULL))
     {
-        if (pipe(pipe_out) < 0)
+        LocalEvent = CreateEventW(NULL, FALSE, FALSE, NULL);
+        if (LocalEvent == NULL)
         {
-            SetLastError(ERROR_OUTOFMEMORY);
-            return 0;
+            Error = GetLastError();
+            goto Quit;
         }
-        fcntl(pipe_out[0], F_SETFD, FD_CLOEXEC);
-        fcntl(pipe_out[1], F_SETFD, FD_CLOEXEC);
+        CreatedEvent = TRUE;
     }
-
-    /* Fork child process */
-    pid = fork();
-    if (pid == -1)
+    else
     {
-        close(pipe_out[0]);
-        close(pipe_out[1]);
-        SetLastError(ERROR_OUTOFMEMORY);
-        return 0;
+        /* Just use whatever the caller gave us */
+        LocalEvent = Event;
+    }
+
+    /* Forward the request to the IP device */
+    Status = NtDeviceIoControlFile(IcmpHandle,
+                                   LocalEvent,
+                                   ApcRoutine,
+                                   ApcContext,
+                                   &IoStatusBlock,
+                                   IOCTL_ECHO_REQUEST,
+                                   IcmpEchoData,
+                                   BufferSize,
+                                   ReplyBuffer,
+                                   ReplySize);
+    if (Status == STATUS_PENDING)
+    {
+        /* Wait */
+        NtWaitForSingleObject(Event, 0, 0);
```

Comment:
You want to wait for LocalEvent here. Or actually, I'd find it more obvious to do `Event = LocalEvent` in the if branch above, then using Event; either works.